File: 2024/rust/src/day1/day1.rs
use std::{collections::HashMap, fs, iter::zip};

fn parse(input: &str) -> (Vec<i32>, Vec<i32>) {
    let parsed = input
        .lines()
        .map(|l| l.split_whitespace().map(|num| num.parse().unwrap()));

    let mut left = Vec::new();
    let mut right = Vec::new();
    for mut e in parsed {
        left.push(e.next().unwrap());
        right.push(e.next().unwrap());
    }
    (left, right)
}

fn puzzle1(input: &str) -> i32 {
    let (mut left, mut right) = parse(input);
    left.sort();
    right.sort();

    zip(left, right).map(|(a, b)| (a - b).abs()).sum()
}

fn puzzle2(input: &str) -> i32 {
    let (left, mut right) = parse(input);
    right.sort();
    let map: HashMap<_, _> = right
        .chunk_by(PartialEq::eq)
        .map(|slice| (slice[0], slice.len() as i32))
        .collect();
    left.iter()
        .map(|&e| e * &map.get(&e).copied().unwrap_or(0))
        .sum()
}

fn main() {
    let input = fs::read_to_string("../inputs/1.txt").unwrap();

    println!("result of puzzle1: {}", puzzle1(&input));
    println!("result of puzzle2: {}", puzzle2(&input));
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_INPUT: &str = r"3   4
4   3
2   5
1   3
3   9
3   3
";

    #[test]
    fn test_puzzle1() {
        assert_eq!(puzzle1(TEST_INPUT), 11);
    }

    #[test]
    fn test_puzzle2() {
        assert_eq!(puzzle2(TEST_INPUT), 31);
    }
}

File: 2024/c++/src/day1/day1.cpp
import std;
import utils;

using namespace assert;


auto parse(std::string_view input) -> std::pair<std::vector<int>, std::vector<int>>
{
    std::stringstream stream{input};
    std::vector<int> left, right;
    int a, b;
    while (stream >> a >> b) {
        left.push_back(a);
        right.push_back(b);
    }
    return {left, right};
}


auto puzzle1(std::string_view input) -> int
{
    auto [left, right] = parse(input);

    std::ranges::sort(left);
    std::ranges::sort(right);

    return std::ranges::fold_left(
        std::views::zip(left, right)
            | std::views::transform([](std::pair<int, int> pair) { return std::abs(pair.first - pair.second); }),
        0,
        std::plus{}
    );
}

auto puzzle2(std::string_view input) -> int
{
    auto [left, right] = parse(input);
    std::ranges::sort(right);
    const auto map = right //
                   | std::views::chunk_by(std::equal_to{})
                   | std::views::transform([](auto x) { return std::pair{*x.begin(), std::ranges::distance(x)}; })
                   | std::ranges::to<std::unordered_map>();
    return std::ranges::fold_left(
        left //
            | std::views::transform([&](const int e) {
                  auto it = map.find(e);
                  return e * (it != map.end() ? it->second : 0);
              }),
        0,
        std::plus{}
    );
}


constexpr std::string_view test_input = R"(3   4
4   3
2   5
1   3
3   9
3   3
)";

auto main() -> int
{
    assert_eq(puzzle1(test_input), 11);
    assert_eq(puzzle2(test_input), 31);

    const std::string input = [] {
        std::ostringstream stream;
        std::ifstream file("../inputs/1.txt");
        stream << file.rdbuf();
        return std::move(stream).str();
    }();
    std::println("result of puzzle1 is: {}", puzzle1(input));
    std::println("result of puzzle2 is: {}", puzzle2(input));
}
